### DISCUSSION TITLE
Best lower-cost machines to run OpenEMS?

### DISCUSSION THREAD

**Discussion by montanaviking** [thliebig/openEMS-Project#195] [General]
I'm currently running E5-2699 V4 and E52690 V4 dual-xeon machines for my computing but these only have about 89GB/sec RAM bandwidth/socket. OK for computations like optimization, but not so great for OpenEMS due to limited RAM bandwidth.
I'd like to look at adding machines which are a better value. While new machines offer great RAM speed, they are very expensive.
I'm considering AMD dual-socket EPYC 7601 which has about double the RAM bandwidth of the above XEON machines/socket due to its 8 RAM channels/socket vs. four RAM channels/socket in the XEONs. EPYC 7551s look attractive too and have 170GB/sec/socket RAM speed.
Yea, the EPYC 9654 would be great at 460GB/sec/socket RAM bandwidth but I don't want to spend $50K on a server.
GPUs would also be a much better value but: 1. memory size limitations (important only for large problems) 2. Not sure about status of OpenEMS on GPUs.
What do you think and what was your choice? 
Thanks!
Phil

**Comment by biergaizi** [comment]
I agree. I'd say that that a dual AMD Zen 2 (EPYC 7000 series) server is the most cost-effective machine in terms of memory bandwidth for individual users for these kinds of simulations. Not too new that it forces you to pay a premium on CPU and RAM, not too old that it lacks 8-channel support. If I were you, I'd make the same decision (I was almost buying one system for my own development, but I eventually decided to use the E5-2680 v4 as all I needed was a low-cost test platform - since I only need to test code for its relative speedups, not to run fast simulations).

But beware that there exists software bottlenecks within openEMS, so it's not currently be able to achieve the full potential of a machine. It may be 50% faster at best, but don't expect magical speedup. NUMA support in openEMS is also currently non-existent, but I still recommend a dual-socket machine, since (1) the price difference between NUMA and non-NUMA is small, (2) you can always run two simulations, each pinned to a socket. (3) Future-proofing. 

I'm working on some optimizations but I do it on my spare time, which you can read about at [1][2][3], so it's moving at a very slow pace. Time and money are not the problem, the limited attention and energy I can spend on this project is. My current plan is to have some things merged into the official openEMS repo before the end of 2024.

[1] https://github.com/thliebig/openEMS-Project/discussions/154#discussioncomment-8183199
[2] https://github.com/thliebig/openEMS/pull/117
[3] https://github.com/thliebig/openEMS-Project/discussions/92

**Reply by montanaviking** [reply]
Thanks biergaizi for the information. And many thanks for your hard work on openEMS. I will look into pinning simulations to a socket. I plan to run several simulations simultaneously for optimization.
When I "retire" next month, if we don't get busy on another project, I will like to look further into openEMS as I will be supporting a small sole proprietor running an RF design company out of his garage to get OpenEMS up and running for circuit board optimizations.
I used OpenEMS extensively for optics and RF in the early 2010s. Should I have time, I'd like to look into the code itself so as to add features and optimizations. I'm especially interested in the GPU versions. 
Thanks again! Phil

**Reply by montanaviking** [reply]
Hi biergaizi,
It turns out I couldn't resist and am getting a four-node 8-socket EPYC 7551. If you need to try things out on it, please kindly let me know. We could connect and I could set up an account for you. I'd like to help your efforts. Will be fun to see what happens.
Best and thanks,
Phil

**Comment by VolkerMuehlhaus** [comment]
> What do you think and what was your choice?

I use openEMS only for work where simulation speed is not an issue. For large models, there is a commercial FDTD implementation that is MUCH faster, so in my opinion it really makes no sense to buy an expensive server for running the slow openEMS solver. Unless of course you need a unique feature like the cylindrical coordinates.

**Reply by montanaviking** [reply]
Thanks VolkerMuehlhaus, for your response.
Yes, I agree that commercial FDTD simulators will be much faster. However, I suspect that the pricing for such software is going to be far north of $10K or more often, you pay for a subscription. Please kindly correct me if I'm wrong. The server I'm getting for this is 4 dual EPYC 7551 nodes for under $3K. This is at least, partially a hobby and yes, it might be cheaper to just rent computer power, e.g. AWS etc... but I like the feeling of owning my own machines and maintaining/configuring them. I should have more time since I'll be "retired" soon and perhaps it could be interesting to see what I could do with openEMS, improving the user interface and/or the compute engine. Thanks, Phil

**Reply by VolkerMuehlhaus** [reply]
Hi Phil, 

yes, cost for fast commercial FDTD is indeed $10k/year or more. I justed wanted to mention the massive difference in solver speed, because accelerating openEMS using high end servers isn't cheap either. That said, your server choice sounds like great value for the money!

Best regards
Volker

**Reply by montanaviking** [reply]
My server finally arrived with 4nodes having 2 EPYC 7551 sockets/node. I'd really like to run MPI on this since the nodes have 25GBE NICs. Wondering if it's possible to pin MPI nodes to sockets or even NUMA nodes? Do MPI nodes only work one MPI node/machine? 
Eventually, I'd like to look into the internals of openEMS' threading model. I think it uses MPI? If so, it would seem to not be too hard to make it NUMA aware since I think MPI has some NUMA tools? I suspect these old EPYCs are quite sensitive to NUMA awareness since they consist of four chips/CPU socket connected internally via AMD's "fabric" which means four NUMA nodes/socket? While the old Intel XEONs have just four RAM channels/socket, they're monolithic with just one NUMA node/socket. I'm thinking that yea, on the surface, old EPYC 7551's should clean up on my XEON E5-2699V4s but it's not that simple since Intel's monolithic construction means fewer NUMA nodes to cross when the software isn't NUMA aware. However, EPYC 7551's might shine if they get the proper NUMA treatment?
Ultimately, I'm envisioning an openEMS version which is initially manually NUMA tunable via a config file then eventually, in later versions, can configure itself via reading the CPU structure.
I have to wonder what the secret of commercial FDTD packages really is? NUMA awareness or tiling? If we did both, could we come close to their performance?
I have a lot of work to do.
Thanks, Phil

**Reply by biergaizi** [reply]
> Eventually, I'd like to look into the internals of openEMS' threading model. I think it uses MPI? If so, it would seem to not be too hard to make it NUMA aware since I think MPI has some NUMA tools?

I don't recommended that, although you can try. 

openEMS has two modes - single-machine mode and multi-machine mode. In single-machine mode, it uses C++ threads (from Boost) and shared memory, so it has implicit communication. The simulation domain is cut across the X axis evenly. In MPI mode, it launches one instance per node with dedicated memory and explicit communication, and on each node it also uses C++ threads. Changing modes requires re-complication because it's controlled by some C macros.

Theoretically, MPI is automatically NUMA-aware if you run one instance per node, but I don't think it's practical to do so - the MPI implementation is only a basic one, lacking communication-avoidance techniques. There's a global lock per timestep for communication, so overhead is very high, speedups only exist for huge problems. This is why the C++ thread mode is the default. But it has the same problem.

This is the motivation of my research on Trapezoid Tiling algorithm - which is a domain decomposition algorithm with native parallelization and communication-avoidance. Hopefully, both modes will become obsolete within a few months after I submit my major rewrite project to the upstream ;-)

**Reply by VolkerMuehlhaus** [reply]
@montanaviking 

> I have to wonder what the secret of commercial FDTD packages really is? NUMA awareness or tiling? If we did both, could we come close to their performance?

We had a brief discussion on that topic here:
https://github.com/thliebig/openEMS-Project/discussions/154#discussioncomment-7857166

I am using both openEMS (for an open source, open PDK project) and commercial FDTD, and compared simulation speed for one of the RFIC inductor models [SG13_L2n0_mesh1.5um_v2.py](https://github.com/IHP-GmbH/IHP-Open-PDK/blob/main/ihp-sg13g2/libs.tech/openems/testcase/SG13_Octagon_L2n0/OpenEMS_Python/SG13_L2n0_mesh1.5um_v2.py).

That model uses PEC boundaries on all sides, so it is somewhat best case for openEMS and avoids the slow PML implementation. For this comparison, I used the exact same mesh and loss model with discretization into skin effect. Model size is 2.1 MCells, 199 x 219 x 49 cells.

I used two machines for testing: 
Desktop with AMD Ryzen 9 5950X (16 Cores) and 32 GB (4x8GB) DDR4 3200 MHz, running Unbuntu 22.04 LTS
Notebook with AMD Ryzen 7 5800H (8 Cores) and 32 GB (2x16GB) DDR4 3200 MHz, running Windows 11 

**Desktop results:**
openEMS simulation time 220s (1231 MCells/s) with number auf threads automatic, used 4 threads
openEMS simulation time 262s (1033 MCells/s) with number auf threads forces to 16 threads
Empire XPU simulation time 18s (11,237 MCells/s) with number auf threads automatic, used 16 threads

**Notebook results**
openEMS simulation time 1305s (205 MCells/s) with number auf threads automatic, used 8 threads
Empire XPU simulation time 52s (3510 MCells/s) with number auf threads automatic, used 8 threads

On the desktop, it was somewhat surprising to see that openEMS was faster using 4 threads than using 16 threads. Only Empire XPU uses all the availables cores efficiently for this testcase.

On the notebook, openEMS selects all available cores, but is more than 20x slower than openEMS whereas it is only 10x on the desktop. Not sure what happens there, that timing can be reproduced and the machine doesn't running other foreground stuff.

<img width="861" alt="L2no_openEMS" src="https://github.com/thliebig/openEMS-Project/assets/47922981/924e270b-802d-4185-853d-fc3aec772c6e">

**Reply by VolkerMuehlhaus** [reply]
One more comment: Above I mentioned the slowdown when using PML boundaries in openEMS, so I compared that as well using the desktop machine with AMD Ryzen 9 5950X (16 Cores) and 32 GB (4x8GB) DDR4 3200 MHz, running Unbuntu 22.04 LTS

When replacing **one** of the six boundaries from PEC to PML_8,
openEMS slows down from **1231** MCells/s to **380** MCells/s
Empire XPU slows down from **11,237** MCells/s to **9840** MCells/s

When replacing **five** of the six boundaries from PEC to PML_8,
openEMS slows down from **1231** MCells/s to **260** MCells/s
Empire XPU slows down from **11,237** MCells/s to **8472** MCells/s

So for models that require good absorbing boundaries (not just the MUR single sheet), openEMS could really benefit from a more efficient PML implementation. In another thread @biergaizi had explained the issue with the current implementation, which really slows down the solver.

**Reply by biergaizi** [reply]
> I have to wonder what the secret of commercial FDTD packages really is? NUMA awareness or tiling?

Multi-timestep tiling is crucial. With perfect tiling code, effective bandwidth is no longer limited by DRAM bandwidth (~50 GB/s) but CPU cache bandwidth. For L3 cache, a ballpark figure is 500 GB/s, which gives 10x performance. In fact, it has been demonstrated in some papers that even L2 cache-oriented tiling is possible in some situations, which gives roughly a 20x speed.

I already gave a detailed bibliography and literature review in the previous post on tiling, pay attention to the DiamondCandy papers - I believe it's currently the state-of-art on FDTD tiling algorithms, but unfortunately I'm not a physicist so I found it's impossible to understand due to lack of solid geometry trainings and intuitions. So I'm using the conventional "trapezoid tiling" solution instead, which does not perform the best but is a classic well-understood solution.

Efficient material encoding likely gives another significant speedup. If the simulation domain only contains a few materials, generating the coefficients on-the-fly would save much bandwidth. Finally, Loop fusion is another trick, Empire XPU probably uses JIT to inline all operations into the main simulation loop, instead of calling each processing step one by one.

>  If we did both, could we come close to their performance?

I'm trying to do both in the past 6 months. In one sentence: it's easier said than done. In particular, tiling algorithms are difficult to design and implement. I found it's extremely difficult to reason anything in 3D space + 1D time.

I've successfully removed the DRAM bandwidth bottleneck. Unfortunately, I think now the CPU starts stalling due to L3 latency and bandwidth limitations due to suboptimal memory access patterns instead, so it just moves the problem around rather than eliminating it. The problem comes from tiling itself - tiling requires accessing non-contiguous and variable-stride data in the middle of the arrays... As a result, I'm only able to achieve around a 1.5x - 2.5x speed from tiling. NUMA-awareness, in comparison, in pretty easy, and the speedup is around 1.4x  - I'm now trying to push that closer to 2x.

This is why my eventual plan is to use GPUs to sidestep the performance gap by brute-force, and to combine GPU with tiling. Hopefully this would give competitive performance but with much lower programming efforts. But GPU code is extremely difficult to debug so my idea is to try improving the CPU code first to test the high-level concepts first.

**Reply by VolkerMuehlhaus** [reply]
> That model uses PEC boundaries on all sides, so it is somewhat best case for openEMS and avoids the slow PML implementation. For this comparison, I used the exact same mesh and loss model with discretization into skin effect. Model size is 2.1 MCells, 199 x 219 x 49 cells.
> 
> I used two machines for testing: Desktop with AMD Ryzen 9 5950X (16 Cores) and 32 GB (4x8GB) DDR4 3200 MHz, running Unbuntu 22.04 LTS Notebook with AMD Ryzen 7 5800H (8 Cores) and 32 GB (2x16GB) DDR4 3200 MHz, running Windows 11
> 
> **Desktop results:** openEMS simulation time 220s (1231 MCells/s) with number auf threads automatic, used 4 threads openEMS simulation time 262s (1033 MCells/s) with number auf threads forces to 16 threads Empire XPU simulation time 18s (11,237 MCells/s) with number auf threads automatic, used 16 threads
> 
> **Notebook results** openEMS simulation time 1305s (205 MCells/s) with number auf threads automatic, used 8 threads Empire XPU simulation time 52s (3510 MCells/s) with number auf threads automatic, used 8 threads

I have tested an additional machine that performs quite well for openEMS: 
HP Z2 Mini G1a workstation with **AMD Ryzen AI Max+ 395** and 128 GB RAM
This CPU has 16 Zen5 cores and twice the memory bandwidth compared to Ryzen 9 9950X

Automatic thread selection picked 7 threads, at an average speed of **1477 Mcells/s**.
Running with numThreads=10, the average speed is **1914 Mcells/s** for this model.

The above numbers are obtained using Ubuntu 24.04. Running openEMS on Windows was notably slower.

This might be a budget-friendly simulation server for openEMS, if you are looking for low power consumption in a small package. The 128 GB RAM used here is surely  overkill for openEMS, but I also need this machine for FEM simulation.

The simulation model moved to [here](https://github.com/IHP-GmbH/IHP-Open-PDK/blob/main/ihp-sg13g2/libs.tech/openems/obsolete/testcase/SG13_Octagon_L2n0/OpenEMS_Python/SG13_L2n0_mesh1.5um_v2.py)

```
Running FDTD engine... this may take a while... grab a cup of coffee?!?
[@        5s] Timestep:         5016 || Speed: 1987.6 MC/s (1.031e-03 s/TS) || Energy: ~4.85e-24 (- 0.00dB)
[@       10s] Timestep:        10032 || Speed: 1879.5 MC/s (1.090e-03 s/TS) || Energy: ~9.97e-22 (- 0.00dB)
[@       15s] Timestep:        15048 || Speed: 2089.8 MC/s (9.801e-04 s/TS) || Energy: ~5.65e-20 (- 0.00dB)
[@       20s] Timestep:        20064 || Speed: 2035.1 MC/s (1.006e-03 s/TS) || Energy: ~1.04e-19 (- 0.00dB)
[@       25s] Timestep:        25080 || Speed: 1941.1 MC/s (1.055e-03 s/TS) || Energy: ~5.73e-18 (- 0.00dB)
[@       31s] Timestep:        30096 || Speed: 1913.0 MC/s (1.071e-03 s/TS) || Energy: ~4.48e-17 (- 0.00dB)
[@       36s] Timestep:        35112 || Speed: 1864.3 MC/s (1.099e-03 s/TS) || Energy: ~1.27e-17 (- 5.48dB)
[@       42s] Timestep:        40128 || Speed: 1899.9 MC/s (1.078e-03 s/TS) || Energy: ~1.27e-16 (- 0.00dB)
[@       47s] Timestep:        45144 || Speed: 1880.6 MC/s (1.089e-03 s/TS) || Energy: ~2.22e-16 (- 0.00dB)
[@       52s] Timestep:        50160 || Speed: 1942.3 MC/s (1.055e-03 s/TS) || Energy: ~4.10e-17 (- 7.34dB)
[@       58s] Timestep:        55176 || Speed: 1938.2 MC/s (1.057e-03 s/TS) || Energy: ~8.94e-19 (-23.94dB)
[@     1m03s] Timestep:        60192 || Speed: 1805.1 MC/s (1.135e-03 s/TS) || Energy: ~6.30e-19 (-25.46dB)
[@     1m09s] Timestep:        65208 || Speed: 1987.3 MC/s (1.031e-03 s/TS) || Energy: ~8.34e-20 (-34.25dB)
[@     1m14s] Timestep:        70224 || Speed: 1920.8 MC/s (1.066e-03 s/TS) || Energy: ~4.33e-20 (-37.10dB)
[@     1m19s] Timestep:        75240 || Speed: 2060.9 MC/s (9.939e-04 s/TS) || Energy: ~2.98e-20 (-38.72dB)
[@     1m24s] Timestep:        80256 || Speed: 2050.4 MC/s (9.990e-04 s/TS) || Energy: ~2.13e-20 (-40.17dB)
[@     1m30s] Timestep:        85272 || Speed: 1840.9 MC/s (1.113e-03 s/TS) || Energy: ~1.57e-20 (-41.50dB)
[@     1m35s] Timestep:        90288 || Speed: 1882.2 MC/s (1.088e-03 s/TS) || Energy: ~1.18e-20 (-42.74dB)
[@     1m40s] Timestep:        95304 || Speed: 1897.2 MC/s (1.080e-03 s/TS) || Energy: ~9.03e-21 (-43.90dB)
[@     1m46s] Timestep:       100320 || Speed: 1862.7 MC/s (1.100e-03 s/TS) || Energy: ~7.00e-21 (-45.01dB)
[@     1m52s] Timestep:       105336 || Speed: 1738.9 MC/s (1.178e-03 s/TS) || Energy: ~5.49e-21 (-46.06dB)
[@     1m57s] Timestep:       110352 || Speed: 1917.8 MC/s (1.068e-03 s/TS) || Energy: ~4.34e-21 (-47.08dB)
[@     2m02s] Timestep:       115368 || Speed: 1940.9 MC/s (1.055e-03 s/TS) || Energy: ~3.46e-21 (-48.06dB)
[@     2m08s] Timestep:       120384 || Speed: 1980.6 MC/s (1.034e-03 s/TS) || Energy: ~2.78e-21 (-49.02dB)
[@     2m13s] Timestep:       125400 || Speed: 1785.2 MC/s (1.147e-03 s/TS) || Energy: ~2.25e-21 (-49.94dB)
[@     2m19s] Timestep:       130416 || Speed: 1840.1 MC/s (1.113e-03 s/TS) || Energy: ~1.82e-21 (-50.85dB)
Time for 130416 iterations with 2048307.00 cells : 139.50 sec
Speed: 1914.89 MCells/s 
```

**Comment by montanaviking** [comment]
Yes indeed, it's expensive any way you look at it. A dream of mine is to
improve open source to greatly reduce this cost to small businesses and
hobbiests. Best, Phil

**Comment by georgy-moshkin** [comment]
I use E2630 CPU based PC connected to my N5095 CPU notebook through USB-to-Ethernet adapter.
I run large simulations on E2630 using Windows remote desktop and small simulations on N5095 notebook in parallel.
It saves a lot of time and keeps notebook responsive. Simulations run around 1.5x faster on E2630 despite having cheapest DDR3 ram.
openEMS reports "Best performance using 6 threads" on E2630 and "3 threads" on N5095.

**Reply by montanaviking** [reply]
Hi georgy-moshkin,
Checking out the Xeon E5-2630, this looks like the processors which are in my old server. It seems that 3 threads/socket ran the fastest for me but that could have been due to NUMA issues?
Also, the E5-2630 has 4-channel RAM whereas the N5095 has 2-channel RAM. Intuitively, it would make sense that a memory-bound code such as openEMS would prefer 3 threads on this processor if the 4-channel RAM processor maximized performance at 6 threads.
When I bought my servers, I aimed for the max number of sockets I could get at a reasonable price. But your results and other things I've heard bring into question the wisdom of throwing more sockets to raise memory bandwidth. It might well be that openEMS lack of NUMA support (I don't mean to criticize openEMS, it's a great code) means one would be better off running it on one socket machines or pinning two running instances of openEMS to their own sockets in a two-socket machine? If so, I'm wondering how one can pin an instance of openEMS to a particular socket? It appears that this would be an optimum throughput strategy when using openEMS as part of an optimization loop as I plan to do in the near future.
Please kindly correct me if I'm wrong and I ask you and other readers to give your opinions. Thanks, Phil

**Reply by georgy-moshkin** [reply]
I am happy that OpenEMS works and gives adequate results. The reason I've bought E2630-based desktop PC is that it supported NVME ssd disk and a lot of cheap RAM. My main goal was to minimize swapping in all possible CAD programs and perform smooth screen recordings for educational videos. I never had swapping problems with OpenEMS, because my models are relatively small.
About your question, I do not have any understanding of NUMA and related stuff. Below is some Windows command that may be worth sharing. I use this command line for screen recordings on Windows:
start /affinity 8 ffmpeg -threads 1 -y -rtbufsize 2048M -f dshow -i video="screen-capture-recorder":audio="virtual-audio-capturer" -r 30 -s 1280x720 -c:v qtrle -preset ultrafast -crf 0 out.mov
(needs this to be present in system: https://github.com/rdp/screen-capture-recorder-to-video-windows-free/releases)
It limits ffmpeg to one core of my CPU.

Note that ffmpeg is started using "start /affinity 8" command where 8 = 0b1000 (fourth bit is set)
/ AFFINITY - Specifies the processor affinity mask as a hexadecimal number. The process is restricted to running on these processors.

Another command line option that may interest you:
/ NODE - Specifies the preferred Non-Uniform Memory Architecture (NUMA) node as a decimal integer.

I think that running two or more instances of OpenEMS in parallel may be beneficial. Each instance can be restricted to certain NUMA node / core using NODE/AFFINITY parameters. I haven't tried it, but my experience with some Microsoft.NET C# computational algorithms (have no idea it was FDTD or something else) was that single EXE run took 20 minutes to complete simulation. Two EXEs executed in parallel took around 30 minutes, which is faster than 20 + 20 = 40 minutes (E2630 CPU). Running three EXEs in parallel had almost no benefit.

There may be some boost with OpenEMS too. E.g., perform parametric sweep in parallel for some model dimension. First simulation L=10 - 0.1 [mm], second simulation L=10 [m], third simulation L=10+0.1[mm]. Need to find out which dimension is the best one. Assume it takes 30+30+30 = 90 minutes if we run next simulation only after previous is finished. The question is can we achieve something like 50 minutes if three of them are running in parallel? I will be very happy if results will be ready 40 minutes earlier than usual. This needs to be tested.

**Reply by montanaviking** [reply]
Hi georgy-moshkin, Please kindly see my latest post "Findings on threads and performance"
Yes, you're right that running jobs in parallel can increase total performance. I'm really interested in knowing if there are any tweaks to openEMS regarding core pinning that would significantly improve performance?
Also, it appears that the older XEONs e.g. E5-2699v4 are faster for single jobs, but the EPYC 7551s are better for multiple parallel jobs when you pin jobs to NUMA nodes. Thanks

**Comment by aaronhance** [comment]
I know this is old, but I believe I have something interesting to add. I have a DDR4 3200Mhz based ryzen 5800h mini PC, Ubuntu, that matches~ the performance of my DDR5 5200Mhz Desktop PC, Ryzen 9950X3D, Windows 11. Both dual channel. Maybe OS is the cause, or perhaps the multi chiplet design hurts here? Perhaps it might be a good idea to make a user benchmark script with system specs and OS, could help with optimisations.

**Reply by VolkerMuehlhaus** [reply]
@aaronhance From my experience, openEMS is indeed faster on Linux when running on the exact same hardware.

**Reply by montanaviking** [reply]
That is interesting. I was looking at OpenParEM as a FEM solver alternative to OpenEMS. However, it appears that OpenEMS, while it's apparently more affected by RAM bandwidth, will also be easier to parameterize for circuit optimizations. While commercial software is much faster, it's also extremely expensive.
In "Findings on threads and performance" posting, I found that of course, pinning the threads to numa nodes yielded much better results for the old EPYC machine -about 2x that of the old XEON E5-2699v4.
Now, regarding Windows, I am developing furnace control software remotely for a small company. I recently changed machines from Windows 10 to Ubuntu 24.04 and compiling is now about 5X faster. Some of this is likely due to the hardware being faster but I've found that Windows file I/O runs about 1/5 the speed of Linux on thumb drives!
I've heard in our Linux group that Windows is notorious for slow file I/O.
I have a question: namely does anyone have a solution they're using to convert GDSII or other layouts to OpenEMS Python formats?
I've been away from OpenEMS for some time but want to revisit.
Thanks

**Reply by VolkerMuehlhaus** [reply]
@montanaviking Regarding your question on GDSII, this might be of interest:
https://github.com/thliebig/openEMS-Project/discussions/371
I also created a similar solution for FEM using AWS Palace.
